### PR TITLE
Fix stack info

### PR DESCRIPTION
**Pull request by vefimenko**
Usage example:
```
Log.Logger = new LoggerConfiguration()
                .WriteTo.StackTraceWrapper(w=>w.Async(x=>x.Airbrake(conf, LogEventLevel.Error)))
                .CreateLogger();
```

### CODE REVIEW

File: StackTraceLogEvent.cs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+                logEvent.MessageTemplate,
+                logEvent.Properties.Select(x => new LogEventProperty(x.Key, x.Value))
+                )
+        {
+            StackTrace = GetStackTrace();
+        }
+
+        private StackTrace GetStackTrace()
+        {
+#if !NETSTANDARD1_4
```
Comment (sergiibratus) [inline StackTraceLogEvent.cs:29]: why do you need this? we are targeting netstandard2 and 2.1

**Review by sergiibratus — COMMENTED** [review]
As I understand the whole problem is that we are too late with the stack trace capturing, so you want to capture it sooner. can you please share where exactly the problem arises and why? I thought it's the airbrake NotifyAsync, but it's not that because the notice is being build in a sync manner. So it should be the Sharpbrake.Serilog sink problem - it allows async Emit calls? or it's smth else?

**Comment by vefimenko** [comment]
> As I understand the whole problem is that we are too late with the stack trace capturing, so you want to capture it sooner. can you please share where exactly the problem arises and why? I thought it's the airbrake NotifyAsync, but it's not that because the notice is being build in a sync manner. So it should be the Sharpbrake.Serilog sink problem - it allows async Emit calls? or it's smth else?

The main problem is in Serilog.Sinks.Async. The Serilog.Sinks.Async [calling our Sink from thread](https://github.com/serilog/serilog-sinks-async/blob/dev/src/Serilog.Sinks.Async/Sinks/Async/BackgroundWorkerSink.cs#L76). Therefore, we cannot get correct call stack in normal way. And calling Sharpbrake synchronously is a bad idea because of network calls.

File: StackTraceLogEvent.cs (hunk 1 above)
Comment (vefimenko) [reply StackTraceLogEvent.cs]: fixed

**Comment by zedle** [comment]
Let me know when this is merged so I can upload the new nupkg

**Comment by zedle** [comment]
Also make sure we bump version to 1.0.6 @vefimenko

File: StackTraceLogEvent.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+            StackTrace = GetStackTrace();
+        }
+
+        private StackTrace GetStackTrace()
+        {
+            var stackTrace = new StackTrace(fNeedFileInfo: true);
+            for (int i = 0; i < stackTrace.FrameCount; i++)
+            {
+                var stackFrame = stackTrace.GetFrame(i);
+                var namesp = stackFrame.GetMethod().DeclaringType?.Namespace;
```
Comment (sergiibratus) [inline StackTraceLogEvent.cs:33]: GetMethod may return null too

File: AirbrakeSink.cs [hunk 3]
```diff
@@ -36,7 +34,23 @@ private Notice BuildNotice(LogEvent logEvent)
             }
             else
             {
-                return _airbrake.BuildNotice(logEvent.RenderMessage());
+                var notice = _airbrake.BuildNotice(logEvent.RenderMessage());
+                if(logEvent is StackTraceLogEvent stackTraceEvent)
+                {
+                    if(notice.Errors.FirstOrDefault() is ErrorEntry error)
+                    {
+                        error.Backtrace = Utils.GetBacktrace(stackTraceEvent.StackTrace);
```
Comment (sergiibratus) [inline AirbrakeSink.cs:42]: is Utils.GetBacktrace missing?

File: AirbrakeLoggerConfigurationExtensions.cs [hunk 4]
```diff
@@ -29,5 +29,19 @@ public static class AirbrakeLoggerConfigurationExtensions
...
+
+        /// <summary>
+        /// Workaround to fix Airbrake call stack for Serilog.Sinks.Async
+        /// </summary>
+        /// <returns></returns>
+        public static LoggerConfiguration StackTraceWrapper(
+            this LoggerSinkConfiguration sinkConfiguration,
+            Action<LoggerSinkConfiguration> configureWrappedSink)
+        {
+            return LoggerSinkConfiguration.Wrap(
```
Comment (sergiibratus) [inline AirbrakeLoggerConfigurationExtensions.cs:41]: nice way to do this! kudos
although this `w=>w.Async(x=>x.Airbrake(conf, LogEventLevel.Error))` looks terrifying :)

**Review by sergiibratus — COMMENTED** [review]
and one general commit - this fixes only half of the problem since ArchonLog in the server does the same, and yeah, this is a different story, just wanted to make sure you will address that too since we are fixing the logs

File: AirbrakeSink.cs (hunk 3 above)
Comment (vefimenko) [reply AirbrakeSink.cs:42]: the class is contained in the Sharpbrake.Client namespace(Sharpbrake library)

File: AirbrakeLoggerConfigurationExtensions.cs [hunk 5]
```diff
@@ -29,5 +29,19 @@ public static class AirbrakeLoggerConfigurationExtensions
         {
             return sinkConfiguration.Sink(new AirbrakeSink(config), restrictedToMinimumLevel);
         }
+
+        /// <summary>
+        /// Workaround to fix Airbrake call stack for Serilog.Sinks.Async
+        /// </summary>
+        /// <returns></returns>
+        public static LoggerConfiguration StackTraceWrapper(
```
[the remark this answers is not in the document]
Comment (vefimenko) [reply AirbrakeLoggerConfigurationExtensions.cs]: renamed to StackTraceCapturer

File: AirbrakeLoggerConfigurationExtensions.cs (hunk 4 above)
Comment (vefimenko) [reply AirbrakeLoggerConfigurationExtensions.cs:41]: any ideas to improve this?

File: StackTraceLogEvent.cs (hunk 2 above)
Comment (vefimenko) [reply StackTraceLogEvent.cs]: fixed

File: StackTraceSink.cs [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    {
+        private readonly ILogEventSink _wrappedSink;
+        public StackTraceSink(ILogEventSink wrappedSink)
+        {
+            _wrappedSink = wrappedSink ?? throw new ArgumentNullException(nameof(wrappedSink));
+        }
+
+        public void Emit(LogEvent logEvent)
+        {
+            _wrappedSink.Emit(new StackTraceLogEvent(logEvent));
```
[the remark this answers is not in the document]
Comment (vefimenko) [reply StackTraceSink.cs]: fixed

**Review by sergiibratus — APPROVED** [review]
LGTM

**Comment by zedle** [comment]
@vefimenko When you're back in can you update the README.md to include the sample usage for the new wrapper

**Pull request merged**